ASoC: SOF: refine suspend_ignored flag

keyonjie opened the pull request:
ASoC: SOF: refine suspend_ignored flag

Move the flag from snd_sof_pcm_stream to snd_sof_dev, use count to store
the number of d0i3 compatible streams being suspended, and then we don't
need the helper snd_sof_dsp_d0i3_on_suspend() anymore.

ranj063 commented:
@keyonjie this PR does exactly the same as #1481 but with a different name for the count. But the second commit in my PR goes one step further in actually simplifying the sequence too. Why is this one better ?

keyonjie commented:
> @keyonjie this PR does exactly the same as #1481 but with a different name for the count. But the 

Yes, it came from yours, sorry about missing your authorship here as I split it from the big local change here, correcting it now. 

> second commit in my PR goes one step further in actually simplifying the sequence too. Why is this one better ?

Yes, your 2nd commit goes further, but I want it to go even further when with S0 D0I3 support enabled, all these are already done on my side, I am refining remained local changes(there might be 5~8 commits)  and will submit them soon.

ranj063 commented:
> > @keyonjie this PR does exactly the same as #1481 but with a different name for the count. But the
> 
> Yes, it came from yours, sorry about missing your authorship here as I split it from the big local change here, correcting it now.
> 
> > second commit in my PR goes one step further in actually simplifying the sequence too. Why is this one better ?
> 
> Yes, your 2nd commit goes further, but I want it to go even further when with S0 D0I3 support enabled, all these are already done on my side, I am refining remained local changes(there might be 5~8 commits) and will submit them soon.

@keyonjie dont worry about authorship. not a big deal, we're the same team. lets put all patches in one PR. It will be hard to see the entire picture if you split them up.

keyonjie commented:
@ranj063 @plbossart OK, I will submit S0 D0i3 support commits here together.

keyonjie commented:
Closing this, it will be included in #1524 soon.

keyonjie commented:
@ranj063 @plbossart S0 D0I3 is implemented now, please help review.

ranj063 commented on sound/soc/sof/sof-priv.h:
```diff
@@ -54,11 +54,17 @@ extern int sof_core_debug;
+/* dsp is in D0I3(low power state) */
+#define	SOF_DSP_STATE_D0I3	(SOF_DSP_STATE_BIT_D0 | SOF_DSP_STATE_BIT_I3)
+/* dsp will update to either D0 or D0I3 state automatically */
+#define	SOF_DSP_STATE_AUTO	SOF_DSP_STATE_BIT_UPDATE
```
@keyonjie this patch is just difficult to understand "ASoC: SOF: refine DSP power states". Care to break it down?

keyonjie replied:
sure, let me move these macro definitions to a separated patch.

ranj063 commented on sound/soc/sof/sof-priv.h:
```diff
@@ -54,11 +54,17 @@ extern int sof_core_debug;
-	SOF_DSP_D0I0 = 0,	/* DSP default D0 substate */
-	SOF_DSP_D0I3,		/* DSP D0i3(low power) substate*/
-};
+#define	SOF_DSP_STATE_BIT_D0		BIT(0) /* dsp is in D0 or D3 */
```
the comment makes no sense. Is the DSP in D0 or D3?

keyonjie replied:
maybe like this?
/* 1 -- dsp is in D0, 0 -- dsp is in D3 */

ranj063 commented on sound/soc/sof/sof-priv.h:
```diff
@@ -54,11 +54,17 @@ extern int sof_core_debug;
+#define	SOF_DSP_STATE_BIT_D0		BIT(0) /* dsp is in D0 or D3 */
+#define	SOF_DSP_STATE_BIT_I3		BIT(1) /* dsp is in I3 or I0 */
+/* dsp power state is under updating */
+#define	SOF_DSP_STATE_BIT_UPDATE	BIT(2)
```
I am not a big fan of bit masks for dsp state. Its not like you could have 2 bits  set at the same time. Enum is a lot more intuitive.

plbossart replied:
Plus don't invent new definitions. I3 does not exist, it's D0i3.

keyonjie replied:
@ranj063 @plbossart let me do this in another way in the coming update version.

ranj063 commented on sound/soc/sof/pm.c:
```diff
@@ -287,7 +264,7 @@ int snd_sof_suspend(struct device *dev)
 		/* suspend to D0i3 */
 		dev_dbg(sdev->dev, "DSP is trying to enter D0i3...\n");
-		ret = snd_sof_set_d0_substate(sdev, SOF_DSP_D0I3);
+		ret = snd_sof_dsp_set_power_state(sdev, SOF_DSP_STATE_D0I3);
```
I think you should move this to snd_sof_dsp_suspend() as well.

ranj063 commented on sound/soc/sof/topology.c:
```diff
@@ -152,15 +153,21 @@ static int sof_keyword_dapm_event(struct snd_soc_dapm_widget *w,
 	spcm = snd_sof_find_spcm_name(scomp, swidget->widget->sname);
 	if (!spcm) {
-		dev_err(scomp->dev, "error: cannot find PCM for %s\n",
+		dev_err(sdev->dev, "error: cannot find PCM for %s\n",
```
please dont change this. let it be scomp->dev

keyonjie replied:
oops, I don't know why I changed this, fix it now.

ranj063 commented on sound/soc/sof/pm.c:
```diff
@@ -258,7 +235,7 @@ int snd_sof_resume(struct device *dev)
 		/* resume from D0I3 */
 		dev_dbg(sdev->dev, "DSP will exit from D0i3...\n");
-		ret = snd_sof_set_d0_substate(sdev, SOF_DSP_D0I0);
+		ret = snd_sof_dsp_set_power_state(sdev, SOF_DSP_STATE_D0);
```
same here. should be moved to snd_sof_dsp_resume()

keyonjie replied:
Yes, actually.

ranj063 commented on sound/soc/sof/intel/hda.c:
```diff
@@ -635,6 +639,8 @@ int hda_dsp_remove(struct snd_sof_dev *sdev)
 #endif
 	hda_codec_i915_exit(sdev);
 
+	hda_dsp_power_state_reset(hda);
```
@keyonjie this looks incorrect. Typically when you remove the SOF device, you resume it and bring it back to full power. So the DSP cant be in D3. 

Besides reset is hardly intuitive to convey the message that the DSP is in D3. please change the name.

keyonjie replied:
> @keyonjie this looks incorrect. Typically when you remove the SOF device, you resume it and bring it back to full power. So the DSP cant be in D3.

Really? Why we need to keep it in full power if it is not used?

> 
> Besides reset is hardly intuitive to convey the message that the DSP is in D3. please change the name.

@ranj063 what name do you suggest to use? My understanding "reset to xxx state" is quite common?

ranj063 replied:
@keyonjie I dont know in my opinion, the mutex isnt required, so just set the power state directly here and remove the init/reset functions

ranj063 commented on sound/soc/sof/intel/hda.h:
```diff
@@ -409,6 +411,8 @@ struct sof_intel_hda_dev {
 
 	/* PM related */
+	u32 dsp_power_state;
+	struct mutex ps_mutex;	/* protects dsp_power_state change */
```
I am also not able to fully appreciate the need for this mutex, at least just yet while reviewing only the second patch of the series. We only set the power states during suspend/resume and they can never overlap

keyonjie replied:
they could be set in IPC sending and PCM open/close also in the subsequent patches.

ranj063 replied:
@keyonjie that was my other comment, why do we need to do it in PCM open/close, can we not do it just when we send the IPC?

kv2019i replied:
@keyonjie This is not a simple lock. In the comment this indicates the lock protects "dsp_power_state", but reading the code, the lock actually covers DSP power state as well. E.g. if you change the DSP power state, you need to hold this lock. Do we really need, or is a lock protecting just the dsp_power_state enough (and if yes, do we need a lock at all)?

ranj063 commented on sound/soc/sof/intel/hda-pcm.c:
```diff
@@ -213,10 +213,10 @@ snd_pcm_uframes_t hda_dsp_pcm_pointer(struct snd_sof_dev *sdev,
 }
 
 int hda_dsp_pcm_open(struct snd_sof_dev *sdev,
-		     struct snd_pcm_substream *substream)
```
@keyonjie which "SOF specific information" do you need easy access to. Neither the patch nor the commit message make it clear

ranj063 commented on sound/soc/sof/intel/hda-dsp.c:
```diff
@@ -446,16 +449,41 @@ int hda_dsp_set_d0i3(struct snd_sof_dev *sdev, u32 flags)
+			     d0i3_work.work);
+	struct snd_sof_dev *sdev = dev_get_drvdata(hda->dev);
+
+	/* keep DMA trace for delay D0I3 */
```
"delay D0i3" you mean D0i3 in S0?

keyonjie replied:
Yes, let me change the comment.

ranj063 commented on sound/soc/sof/intel/hda-pcm.c:
```diff
@@ -227,12 +228,21 @@ int hda_dsp_pcm_open(struct snd_sof_dev *sdev,
+	if (sstream->d0i3_compatible)
+		hda->lp_pcm_active++;
+
+	snd_sof_dsp_set_power_state(sdev, SOF_DSP_STATE_AUTO);
```
Do you really need this here? You can do this after IPC TX no?

keyonjie replied:
Here is PCM open, Do you mean doing it after trigger start IPC, in pcm_trigger?

ranj063 commented on sound/soc/sof/intel/hda.c:
```diff
@@ -597,6 +600,9 @@ int hda_dsp_remove(struct snd_sof_dev *sdev)
 	const struct sof_intel_dsp_desc *chip = hda->desc;
 
+	/* First Cancel any pending attempt to put DSP to D0i3 */
+	cancel_delayed_work_sync(&hda->d0i3_work);
```
don't you also need to cancel this work during system suspend?

keyonjie replied:
you mean D0I3 suspend? that is already covered by hda_dsp_suspend()->hda_dsp_set_power_state() here:
```
	case SOF_DSP_STATE_D0I3:
		/* First Cancel any pending attempt to put DSP to D0i3 */
		cancel_delayed_work_sync(&hda->d0i3_work);

		/* disable DMA trace for no delay D0I3 */
		ret = hda_dsp_set_d0i3(sdev, HDA_PM_NO_DMA_TRACE);

```

ranj063 commented on sound/soc/sof/intel/hda-dsp.c:
```diff
@@ -338,49 +338,154 @@ static int hda_dsp_send_pm_gate_ipc(struct snd_sof_dev *sdev, u32 flags)
 	return ret;
 }
 
+int hda_dsp_set_d0i3(struct snd_sof_dev *sdev, u32 flags)
```
There's too  much in common between set D0i0 and set_D0i3. I think we can unify these 2 functions.

keyonjie replied:
um, 3 biggest differences that why I split them, the 1st is that we only need delayed work for set D0I3(not for set D0I0), the 2nd one is that we need opposite sequence of IPC sending and register setting, and the 3rd one is that we need vary POWER_GATE IPC flags for D0I3 only.

ranj063 commented on sound/soc/sof/intel/hda-dsp.c:
```diff
@@ -334,8 +334,9 @@ static int hda_dsp_send_pm_gate_ipc(struct snd_sof_dev *sdev, u32 flags)
 	/* send pm_gate ipc to dsp */
-	return sof_ipc_tx_message(sdev->ipc, pm_gate.hdr.cmd, &pm_gate,
-				  sizeof(pm_gate), &reply, sizeof(reply));
+	return sof_ipc_tx_message_nopm(sdev->ipc, pm_gate.hdr.cmd,
```
is this change really necessary? You'd always send the pm_date_ipc before the DSp is put in D0i3, isnt it?

keyonjie replied:
It is needed, without this x_nopm() helpers, it will wake up the DSP to D0I0 before sending the pm_gate_ipc.

keyonjie commented:
WoW, looks this will become another 200+ comments PR easily, if I reply them one by one, :).

I prefer to update to a new version firstly next week when I am back to Shanghai.

plbossart commented:
> WoW, looks this will become another 200+ comments PR easily, if I reply them one by one, :).
> 
> I prefer to update to a new version firstly next week when I am back to Shanghai.

I don't mean to sound rude but if you have 200+ comments you probably need to recheck this PR on your own and resubmit when it's mature.

keyonjie commented:
> > WoW, looks this will become another 200+ comments PR easily, if I reply them one by one, :).
> > I prefer to update to a new version firstly next week when I am back to Shanghai.
> 
> I don't mean to sound rude but if you have 200+ comments you probably need to recheck this PR on your own and resubmit when it's mature.

That's true, thanks for reminding.

keyonjie commented:
@plbossart @ranj063 Happy thanks giving day!

I think I addressed most of the comments in the updated version, hope you like this version.

I talked with @RanderWang and realize that we might need to change to put Audio DSP to D0I3 for runtime suspend, but I think there is no conflict with what I implemented here, though it might be possible to combine these 2 scenarios with change in ASoC PCM handling(I did this try at the very beginning of my start of the D0Ix task in https://github.com/thesofproject/linux/pull/1154), some people might not like this solution as it change the component device RT PM logic in soc-dpcm.c when most of their platform actually don't need that extra logic.

ranj063 commented on sound/soc/sof/intel/hda-dsp.c:
```diff
@@ -338,49 +338,155 @@ static int hda_dsp_send_pm_gate_ipc(struct snd_sof_dev *sdev, u32 flags)
+exit:
+	mutex_unlock(&hda->ps_mutex);
+	return ret;
+}
```
@keyonjie I'm going to leave a long comment about the changes in this commit.

1. The checks for power state inside the set_D0i0/set_D0i3 are a bit of an overkill. In reality, there are only 2 options: enter D0i3 or exit D0i3. Entering is always from D0, you'd never really see the case where you're trying to enter D0i3 from D3. Likewise, exiting D0i3 is only possible if you're already in D0i3. Otherwise, it would be a system resume or a runtime resume. 

2. I still do not appreciate the mutex. The commands to enter/exit D0i3 are already protected by waiting for the CIP timeout. Do we really need a mutex for just setting the hda->dsp_power_state? This is not really a register write. Maybe @kv2019i and @plbossart can correct me if I am wrong

3. The bit masks for the power states are also not pretty. There are 3 states, why cant we just simply use an enum instead of setting a D0 bit along with I3 bit? Makes it a lot easier to understand

keyonjie replied:
> @keyonjie I'm going to leave a long comment about the changes in this commit.
> 
> 1. The checks for power state inside the set_D0i0/set_D0i3 are a bit of an overkill. In reality, there are only 2 options: enter D0i3 or exit D0i3. Entering is always from D0, you'd never really see the case where you're trying to enter D0i3 from D3. Likewise, exiting D0i3 is only possible if you're already in D0i3. Otherwise, it would be a system resume or a runtime resume.

That's true, but since the enum dsp_power_state can have the D3 value, we only implement the helpers here, we can't make sure the caller will use it correctly always?

Safe check is a compromise, some scenarios might never happen(e.g. D3->D0I3 entry, D3->D0I3 exit), some might happen via error used(e.g. D0I3->D0I3 entry, D0->D0I3 exit).

I preferred to add these check here to avoid critical errors happen or disordering the state(e.g. if we do try doing D3->D0I3), but I am fine if both you and @plbossart think we should remove the D3 check.

> 2. I still do not appreciate the mutex. The commands to enter/exit D0i3 are already protected by waiting for the CIP timeout. Do we really need a mutex for just setting the hda->dsp_power_state? This is not really a register write. Maybe @kv2019i and @plbossart can correct me if I am wrong

I introduced the mutex as I think we could have race in setting and getting this dsp_power_state, especially considering that the hda_dsp_set_d0i3() could be a delayed work which could run in another thread.

> 3. The bit masks for the power states are also not pretty. There are 3 states, why cant we just simply use an enum instead of setting a D0 bit along with I3 bit? Makes it a lot easier to understand

The key point here is that sometimes the caller(e.g. after a normal IPC sent) actually don't know which state it should put the DSP to, I introduce a command SOF_DSP_STATE_SET_AUTO here for these scenarios. So the rule here is only using command when calling hda_dsp_set_power_state(), and leave the DSP state change decision to the dsp state management logic(inside hda-dsp.c) only.

```
enum sof_dsp_state {
	SOF_DSP_STATE_D3 = 0,
	SOF_DSP_STATE_D0 = SOF_DSP_STATE_BIT_D0,
	SOF_DSP_STATE_D0I3 = SOF_DSP_STATE_BIT_D0 | SOF_DSP_STATE_BIT_I3,
};

enum sof_dsp_state_cmd {
	SOF_DSP_STATE_SET_D3 = SOF_DSP_STATE_D3,
	SOF_DSP_STATE_SET_D0 = SOF_DSP_STATE_D0,
	SOF_DSP_STATE_SET_D0I3 = SOF_DSP_STATE_D0I3,
	SOF_DSP_STATE_SET_AUTO = SOF_DSP_STATE_BIT_UPDATE,
};
```

plbossart commented on sound/soc/sof/pm.c:
```diff
@@ -255,7 +255,7 @@ int snd_sof_resume(struct device *dev)
 	int ret;
 
-	if (snd_sof_dsp_d0i3_on_suspend(sdev)) {
+	if (sdev->d0i3_suspend_cnt) {
```
This direction is not quite right.
It turns out that for SoundWire we do have to keep the DSP powered if we want to avoid complete re-enumeration when exiting clock stop. This could happen in the complete absence of 'D0i3 compatible' streams.
this is really missing an idle loop detection that is based on 
a) D0i3-compatible streams being opened
b) no IPC for a while
c) the state of child devices, e.g. if a SoundWire master remains 'active' preventing pm_runtime entry.

ranj063 replied:
@plbossart I didnt quite understand this ask. The change here only really affects system suspend when there are D0i3-compatible streams running in the DSP. Are you saying that we also need to account for the fact that the SDW master could be active?

plbossart replied:
what I am saying is that while doing a system suspend, the device may already be in D0i3 and remain there (S0 target) or move to D3 (S3 target).

The use of refcounts would be quite misleading here, it's only one of the reasons why the device might remain in D0i3.

ranj063 replied:
@plbossart agreed wrt the WoV usecase this is the only reason that the DSP would remain in D0i3. Maybe we're missing the line of sight for the other possible indicators for this. Can we possibly think of adding this in a separate PR?

plbossart replied:
well no. You remove a criterion to replace it by a refcount, we are not going to modify this again later to come back to another criterion. We need to respect the time of reviewers...

keyonjie replied:
@plbossart I think I get your point here, I talked with @RanderWang about keeping the DSP in D0I3 for runtime suspend (for SDW clock stop supporting), I did't intend mix that with the D0I3-compatible stream running on D0I3, as I commented in https://github.com/thesofproject/linux/pull/1524#issuecomment-559761166.

keyonjie replied:
Hi @plbossart @ranj063 so the state machine could be updated like this:

```
/*
 * Audio DSP states may transform as below:+
 *
 *                +---------------------+  runtime resume
 *                |                     <---------------------+
 *   +------------>                     |  runtime suspend    |
 *   |            |       D0(active)    +----------------+    |
 *   |            |                     +------------+   |    |
 *   |       +----+                     |    D0I3    |   |    |
 *   |       |    +-------^----+----^-- +  compatible|   |    |
 *   |       |            |    |    |        stream  |   |    |
 *   |resume |            |    |    |        opened  |   |    |
 *   | from  |  D3        |    |    |        only    |   |    |
 *   |  D3   |suspend     |    |    |                |   |    |
 *   |       |            |    |    |The last        |   |    |
 *   |       |            |    |    |opened D0i3     |   |    |
 *   |       |            |    |    |compatible      |   |    |
 *   |       |            |    |    |stream closed   |   |    |
 * +-+-------v-------+    |    |    |            +---v---v----+---+
 * |                 |    |    |    +------------+                |
 * |  D3 (suspended) |    |    | D0I3 suspend    |      D0I3      |
 * |                 |    |    +----------------->                +---+
 * |                 |    +----------------------+                |   |
 * +-------^---------+         resume from       +----------------+   |
 *         |                   D0I3 suspend                           |
 *         |                                   D3 suspend             |
 *         +----------------------------------------------------------+
 *
 * d0i3_suspend = s0_suspend && D0I3 stream opened,
 * D3 suspend = !d0i3_suspend,
 */

```

ranj063 replied:
@keyonjie are you saying that you'd never enter D3 during runtime suspend then>?

keyonjie replied:
Yes, @ranj063 enter to D0I3 only for runtime suspend, if the platform doesn't support D0I3, the power gating will still work when it is not used, for power saving.

plbossart replied:
NAK, runtime suspend and D0i3 are mutually incompatible.

keyonjie replied:
@plbossart can you help guide how the state machine should look like in your brain?

kv2019i commented on sound/soc/sof/pm.c:
```diff
@@ -284,7 +284,7 @@ int snd_sof_suspend(struct device *dev)
 	int ret;
 
-	if (snd_sof_dsp_d0i3_on_suspend(sdev)) {
+	if (sdev->d0i3_suspend_cnt) {
```
Hey, I wasn't on the review list, so I'm a bit late to this. I'm not thrilled about adding a refcount, especially as the places in code where it is incremented and decremented, have a lot of additional  conditional logic of their own. When/if something goes wrong, it seems painful to debug where the refcount went wrong. With old snd_sof_dsp_d0i3_on_suspend(), you at least have a single place to check which stream is in unexpected state. And a place to add new logic that can keep DSP at D0i3.

keyonjie replied:
> Late to the review, but a few specific comments (on new locks and refcounts), please see inline.

Thanks for review @kv2019i, it comes in time we are still in discussion and I just get myself free from other higher priority issues.

keyonjie replied:
> Hey, I wasn't on the review list, so I'm a bit late to this. I'm not thrilled about adding a refcount, especially as the places in code where it is incremented and decremented, have a lot of additional conditional logic of their own. When/if something goes wrong, it seems painful to debug where the refcount went wrong. With old snd_sof_dsp_d0i3_on_suspend(), you at least have a single place to check which stream is in unexpected state. And a place to add new logic that can keep DSP at D0i3.

Yes, @plbossart also mentioned about this, my opinion is that if the criteria we can foresee is only that "the count of D0I3 compatible streams being suspend > 0", I agree with @ranj063 that using this simple refcount is fine, but if we can foresee that we need to add some other criteria (e.g. for SDW) for D0I3 entry, yes, I agree with you and @plbossart here.

keyonjie commented:
> @keyonjie This is not a simple lock. In the comment this indicates the lock protects "dsp_power_state", but reading the code, the lock actually covers DSP power state as well. E.g. if you change the DSP power state, you need to hold this lock. Do we really need, or is a lock protecting just the dsp_power_state enough (and if yes, do we need a lock at all)?

Yes, the intention is to protect the whold DSP power state change process.

keyonjie commented:
> Late to the review, but a few specific comments (on new locks and refcounts), please see inline.

Thanks for review @kv2019i, it comes in time we are still in discussion and I just get myself free from other higher priority issues.

keyonjie commented:
@plbossart @ranj063 @kv2019i the PR is updated with removing adding the d0i3_suspend_cnt(and keep using the helper snd_sof_dsp_d0i3_on_suspend() which can be extended for SDW later), I keep the usage of the mutex to protect the DSP power state access *and the state changing itself* as the state change can take long time during the delayed_work, e.g. if we forget to cancel the delayed work before trying to clear/reset the state, this mutex will help to reject this kind of operation.

plbossart commented:
> @plbossart @ranj063 @kv2019i the PR is updated with removing adding the d0i3_suspend_cnt(and keep using the helper snd_sof_dsp_d0i3_on_suspend() which can be extended for SDW later), I keep the usage of the mutex to protect the DSP power state access _and the state changing itself_ as the state change can take long time during the delayed_work, e.g. if we forget to cancel the delayed work before trying to clear/reset the state, this mutex will help to reject this kind of operation.

@keyonjie I am not going to review 2 PRs for the same things. Please work with @ranj063 and suggest a common solution.

keyonjie commented:
> @keyonjie I am not going to review 2 PRs for the same things. Please work with @ranj063 and suggest a common solution.

Sorry for the confusion, yes, please hold on for a moment, will let you know once we are aligned.

keyonjie commented:
#1631 will be the replacement for this, closing this one.
